tidb-binlog/driver/reader/kafka_consumer/sarama.go

```diff
@@ -0,0 +1,260 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafka_consumer
+
+import (
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+)
+
+const (
+	saramaType = "sarama"
+)
+
+type Sarama struct {
+	client    sarama.Client
+	consumer  sarama.Consumer
+	topic     string
+	partition int32
+}
+
+func NewSaramaConsumer(cfg *KafkaConfig) (Consumer, error) {
+	conf := sarama.NewConfig()
+	// set to 10 minutes to prevent i/o timeout when reading huge message
+	conf.Net.ReadTimeout = KafkaWaitTimeout
+
+	// open error
+	conf.Consumer.Return.Errors = true
+	if cfg.SaramaBufferSize > 0 {
+		conf.ChannelBufferSize = cfg.SaramaBufferSize
+	}
+
+	client, err := sarama.NewClient(cfg.Addr, conf)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	consumer, err := sarama.NewConsumerFromClient(client)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &Sarama{
+		client:    client,
+		consumer:  consumer,
+		topic:     cfg.Topic,
+		partition: cfg.Partition,
+	}, nil
+}
+
+// ConsumeFromOffset implements kafka.Consumer.ConsumerFromOffset
+func (s *Sarama) ConsumeFromOffset(offset int64, consumerChan chan<- *KafkaMsg, done <-chan struct{}) error {
+	log.Info("start consumer on",
+		zap.String("topic", s.topic),
+		zap.Int32("partition", s.partition),
+		zap.Int64("offset", offset))
+	partitionConsumer, err := s.consumer.ConsumePartition(s.topic, s.partition, offset)
+	if err != nil {
+		return errors.Trace(err)
+
+	}
+	defer partitionConsumer.Close()
+
+	for {
+		select {
+		case <-done:
+			log.Info("finish consumer")
+			return nil
+
+		case emsg := <-partitionConsumer.Errors():
+			log.Error("consume partitionConsumer failed:",
+				zap.String("topic", emsg.Topic),
+				zap.Int32("partitiong", emsg.Partition),
+				zap.Error(emsg.Err))
+			continue
+
+		case kmsg, ok := <-partitionConsumer.Messages():
+			if !ok {
```

Comment:
if !ok means the partitionConsumer is closed, it will never became ok here

Reply:
remove !ok [branch](https://github.com/pingcap/tidb-tools/pull/292/files#diff-8de333e66d7d8d94c619760775d51273R89)